src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,560 @@
+         (segment == 7 && digits <= 2) ||
+         (segment != 0 && segment != 6 && segment != 7 && digits > 0 && digits <= 2);
+}
+
+/**
+ * Ported from Spark:
+ *   https://github.com/apache/spark/blob/v3.5.0/sql/api/src/main/scala/
+ *   org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L394
+ *
+ * Parse a string with time zone to a timestamp.
+ * The bool in the returned tuple is false if the parse failed.
+ */
+__device__ __host__ thrust::tuple<cudf::timestamp_us, bool> parse_string_to_timestamp_us(
+  cudf::string_view const& timestamp_str,
+  const char* default_time_zone,
+  cudf::size_type default_time_zone_char_len,
+  bool allow_time_zone,
+  bool allow_special_expressions,
+  cudf::timestamp_us epoch,
+  cudf::timestamp_us now,
+  cudf::timestamp_us today,
+  cudf::timestamp_us tomorrow,
+  cudf::timestamp_us yesterday)
+{
+  auto error_us = cudf::timestamp_us{cudf::duration_us{0}};
+
+  if (timestamp_str.empty()) { return thrust::make_tuple(error_us, false); }
+
+  const char* curr_ptr = timestamp_str.data();
+  const char* end_ptr  = curr_ptr + timestamp_str.size_bytes();
+
+  // trim left
+  while (curr_ptr < end_ptr && is_whitespace(*curr_ptr)) {
+    ++curr_ptr;
+  }
+  // trim right
+  while (curr_ptr < end_ptr - 1 && is_whitespace(*(end_ptr - 1))) {
+    --end_ptr;
+  }
+
+  // special strings: epoch, now, today, yesterday, tomorrow
+  if (allow_special_expressions) {
+    char const* begin_epoch = "epoch";
+    char const* end_epoch   = begin_epoch + 5;
+
+    char const* begin_now = "now";
+    char const* end_now   = begin_now + 3;
+
+    char const* begin_today = "today";
+    char const* end_today   = begin_today + 5;
+
+    char const* begin_tomorrow = "tomorrow";
+    char const* end_tomorrow   = begin_tomorrow + 8;
+
+    char const* begin_yesterday = "yesterday";
+    char const* end_yesterday   = begin_yesterday + 9;
+
+    if (equals_ascii_ignore_case(curr_ptr, end_ptr, begin_epoch, end_epoch)) {
+      // epoch
+      return thrust::make_tuple(epoch, true);
+    } else if (equals_ascii_ignore_case(curr_ptr, end_ptr, begin_now, end_now)) {
+      // now
+      return thrust::make_tuple(now, true);
+    } else if (equals_ascii_ignore_case(curr_ptr, end_ptr, begin_today, end_today)) {
+      // today
+      return thrust::make_tuple(today, true);
+    } else if (equals_ascii_ignore_case(curr_ptr, end_ptr, begin_tomorrow, end_tomorrow)) {
+      // tomorrow
+      return thrust::make_tuple(tomorrow, true);
+    } else if (equals_ascii_ignore_case(curr_ptr, end_ptr, begin_yesterday, end_yesterday)) {
+      // yesterday
+      return thrust::make_tuple(yesterday, true);
+    }
+  }
+
+  if (curr_ptr == end_ptr) { return thrust::make_tuple(error_us, false); }
+
+  const char* const bytes   = curr_ptr;
+  const size_t bytes_length = end_ptr - curr_ptr;
+
+  thrust::optional<cudf::string_view> tz;
+  int segments[]             = {1, 1, 1, 0, 0, 0, 0, 0, 0};
```

Comment:
nit: Some comments about the segments might be helpful. Spark does not have any comments, but the code is hard to follow.

Reply:
Done.
Added:
```java
    // segments stores: [year, month, day, hour, minute, seconds, microseconds, no_use_item, no_use_item]
    // the two tail items are no use, but here keeps them as Spark does
    int segments[]             = {1, 1, 1, 0, 0, 0, 0, 0, 0};

```